core/vm/exit_code/exit_code.hpp

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef CPP_FILECOIN_CORE_VM_EXITCODE_EXITCODE_HPP
+#define CPP_FILECOIN_CORE_VM_EXITCODE_EXITCODE_HPP
+
+#include "boost/variant.hpp"
+
+namespace fc::vm::exit_code {
+
+  /**
+   * @brief Successful exit code
+   */
+  struct Success {
+    /**
+     * to be EqualityComparable, always return true
+     * @param other - other Success object
+     * @return true
+     */
+    bool operator==(const Success &other) const;
+  };
+
+  /**
+   * @brief System error codes
+   */
+  enum class SystemError {
+    // Represents a failure to find an actor.
+    kActorNotFound = 1,
+
+    // Represents a failure to find the code for a particular actor in the VM
+    // registry.
+    kActorCodeNotFound,
+
+    // Represents a failure to find a method in an actor
+    kInvalidMethod,
+
+    // Indicates that a method was called with the incorrect number of
+    // arguments, or that its arguments did not satisfy its preconditions
+    kInvalidArgumentsSystem,
+
+    // Represents a failure to apply a message, as it did not carry sufficient
+    // funds for its application
+    kInsufficientFundsSystem,
+
+    // Represents a message invocation out of sequence. This happens when
+    // message.CallSeqNum is not exactly actor.CallSeqNum + 1
+    kInvalidCallSeqNum,
+
+    // Returned when the execution of an actor method (including its subcalls)
+    // uses more gas than initially allocated.
+    kOutOfGas,
+
+    // Returned when an actor method invocation makes a call to the runtime that
+    // does not satisfy its preconditions.
+    kRuntimeAPIError,
+
+    // Returned when an actor method invocation calls rt.Assert with a false
+    // condition.
+    kRuntimeAssertFailure,
+
+    // Returned when an actor method's Send call has returned with a failure
+    // error code (and the Send call did not specify to ignore errors).
+    kMethodSubcallError
+  };
+
+  /**
+   * @brief User defined error codes
+   */
+  enum class UserDefinedError {
+    kInsufficientFundsUser = 1,
+    kInvalidArgumentsUser,
+    kInconsistentStateUser,
+
+    kInvalidSectorPacking,
+    kSealVerificationFailed,
+    kPoStVerificationFailed,
+    kDeadlineExceeded,
+    kInsufficientPledgeCollateral,
+  };
+
+  /**
+   * @brief Virtual machine exit code
+   */
+  class ExitCode {
+   public:
+    using TValue = boost::variant<Success, SystemError, UserDefinedError>;
```

Comment:
It does not look like the right solution for error handling. Better to have a single type (or a common parent) for the error case. The reason is - when TValue is received, then you have to employ two visitors instead of a single for error processing. The current approach is less maintainable and performant at runtime.

Reply:
reworked